File: kotlin/src/test/kotlin/com/bhengubv/circleai/brain/RagTest.kt
// RagTest.kt
//
// Exercises RagContextBuilder + RagPipelineBuilder. Mirrors the verified
// TypeScript suite (tests/rag.test.ts) and C# RagContextBuilderTests plus the
// fluent-builder surface and the embedder ranking path.

package com.bhengubv.circleai.brain

import com.bhengubv.circleai.memory.brain.EpisodicEntry
import com.bhengubv.circleai.memory.brain.IEpisodicStore
import com.bhengubv.circleai.memory.brain.ITextEmbedder
import com.bhengubv.circleai.memory.brain.InMemoryEpisodicStore
import com.bhengubv.circleai.memory.brain.RagContextBuilder
import com.bhengubv.circleai.memory.brain.RagPipelineBuilder
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Test
import java.time.Instant
import java.util.UUID
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

class RagTest {

    private fun episodic(
        id: String = UUID.randomUUID().toString(),
        userText: String = "u",
        assistantText: String = "a",
        recordedAtUtc: Instant = Instant.parse("2026-06-01T12:34:00Z"),
        appContext: String? = null,
        embedding: FloatArray? = null,
    ): EpisodicEntry = EpisodicEntry(
        id = id,
        userText = userText,
        assistantText = assistantText,
        recordedAtUtc = recordedAtUtc,
        appContext = appContext,
        embedding = embedding,
    )

    private fun countOccurrences(text: String, token: String): Int {
        var count = 0
        var start = 0
        while (true) {
            val i = text.indexOf(token, start)
            if (i < 0) break
            count++
            start = i + token.length
        }
        return count
    }

    /** Store that always throws — used to test resilience. */
    private class ThrowingEpisodicStore : IEpisodicStore {
        override suspend fun addAsync(entry: EpisodicEntry): Unit = throw RuntimeException("store failure")
        override suspend fun searchAsync(queryEmbedding: FloatArray?, topK: Int): List<EpisodicEntry> =
            throw RuntimeException("store failure")
        override suspend fun getRecentAsync(count: Int): List<EpisodicEntry> = throw RuntimeException("store failure")
        override suspend fun countAsync(): Int = throw RuntimeException("store failure")
        override suspend fun pruneOlderThanAsync(cutoff: Instant): Int = throw RuntimeException("store failure")
    }

    // ══════════════════════════════════════════════════════════════════════
    // Empty / missing query
    // ══════════════════════════════════════════════════════════════════════

    @Test
    fun `empty query returns empty`() = runTest {
        val b = RagContextBuilder(InMemoryEpisodicStore())
        assertEquals("", b.buildContextAsync(""))
    }

    @Test
    fun `whitespace query returns empty`() = runTest {
        val b = RagContextBuilder(InMemoryEpisodicStore())
        assertEquals("", b.buildContextAsync("   "))
    }

    // ══════════════════════════════════════════════════════════════════════
    // Empty store
    // ══════════════════════════════════════════════════════════════════════

    @Test
    fun `empty store returns empty`() = runTest {
        val b = RagContextBuilder(InMemoryEpisodicStore())
        assertEquals("", b.buildContextAsync("hello"))
    }

    // ══════════════════════════════════════════════════════════════════════
    // Non-empty store — recency fallback (no embedder)
    // ══════════════════════════════════════════════════════════════════════

    @Test
    fun `returns a formatted block with the header and both texts`() = runTest {
        val store = InMemoryEpisodicStore()
        store.addAsync(
            episodic(
                userText = "What is SDPKT?",
                assistantText = "SDPKT is the TGN wallet.",
                recordedAtUtc = Instant.parse("2026-06-01T11:00:00Z"),
            ),
        )

        val b = RagContextBuilder(store, null, 3)
        val result = b.buildContextAsync("tell me about the wallet")

        assertNotEquals("", result)
        assertTrue(result.contains("What is SDPKT?"))
        assertTrue(result.contains("SDPKT is the TGN wallet."))
        assertTrue(result.contains("[Relevant past exchanges"))
    }

    @Test
    fun `formats the UTC timestamp as yyyy-MM-dd HH mm and labels User B`() = runTest {
        val store = InMemoryEpisodicStore()
        store.addAsync(episodic(userText = "q", assistantText = "r", recordedAtUtc = Instant.parse("2026-06-01T09:05:00Z")))
        val b = RagContextBuilder(store, null, 1)
        val result = b.buildContextAsync("anything")
        assertTrue(result.contains("[2026-06-01 09:05 UTC]"))
        assertTrue(result.contains("User: q"))
        assertTrue(result.contains("B!: r"))
    }

    @Test
    fun `respects topK (counts bullet prefixes)`() = runTest {
        val store = InMemoryEpisodicStore()
        repeat(10) { i -> store.addAsync(episodic(userText = "question $i", assistantText = "answer $i")) }

        val b = RagContextBuilder(store, null, 2)
        val result = b.buildContextAsync("any question")
        assertEquals(2, countOccurrences(result, "• ["))
    }

    @Test
    fun `includes the app context when set`() = runTest {
        val store = InMemoryEpisodicStore()
        store.addAsync(episodic(userText = "bid query", assistantText = "bid answer", appContext = "tgn.bidbaas"))
        val b = RagContextBuilder(store, null, 3)
        val result = b.buildContextAsync("bidding")
        assertTrue(result.contains("tgn.bidbaas"))
    }

    @Test
    fun `truncates long texts to half the per-entry budget with an ellipsis`() = runTest {
        val store = InMemoryEpisodicStore()
        val longText = "x".repeat(500)
        store.addAsync(episodic(userText = longText, assistantText = "a"))
        // maxCharsPerEntry 100 → half 50 → truncate to 49 chars + "…"
        val b = RagContextBuilder(store, null, 1, 100)
        val result = b.buildContextAsync("q")
        assertTrue(result.contains("x".repeat(49) + "…"))
        assertTrue(!result.contains("x".repeat(51)))
    }

    // ══════════════════════════════════════════════════════════════════════
    // Embedder ranking path
    // ══════════════════════════════════════════════════════════════════════

    @Test
    fun `ranks by the embedding when an embedder is supplied`() = runTest {
        val store = InMemoryEpisodicStore()
        store.addAsync(episodic(userText = "near", assistantText = "n", embedding = floatArrayOf(1f, 0f)))
        store.addAsync(episodic(userText = "far", assistantText = "f", embedding = floatArrayOf(0f, 1f)))

        // Embedder maps any query to the x-axis, so "near" should rank first.
        val embedder = object : ITextEmbedder {
            override suspend fun generateAsync(text: String): FloatArray = floatArrayOf(1f, 0f)
        }
        val b = RagContextBuilder(store, embedder, 1)
        val result = b.buildContextAsync("anything")
        assertTrue(result.contains("near"))
        assertTrue(!result.contains("far"))
    }

    @Test
    fun `falls back to recency when the embedder throws (still best-effort)`() = runTest {
        val store = InMemoryEpisodicStore()
        store.addAsync(
            episodic(userText = "only", assistantText = "entry", recordedAtUtc = Instant.parse("2026-06-01T00:00:00Z")),
        )
        val embedder = object : ITextEmbedder {
            override suspend fun generateAsync(text: String): FloatArray = throw RuntimeException("embedder offline")
        }
        val b = RagContextBuilder(store, embedder, 3)
        val result = b.buildContextAsync("q")
        assertTrue(result.contains("only"))
    }

    // ══════════════════════════════════════════════════════════════════════
    // Resilience — store throws
    // ══════════════════════════════════════════════════════════════════════

    @Test
    fun `returns empty when the store throws (RAG is best-effort)`() = runTest {
        val b = RagContextBuilder(ThrowingEpisodicStore())
        assertEquals("", b.buildContextAsync("query"))
    }

    // ══════════════════════════════════════════════════════════════════════
    // RagPipelineBuilder
    // ══════════════════════════════════════════════════════════════════════

    @Test
    fun `builds from an in-memory store and produces a working builder`() = runTest {
        val store = InMemoryEpisodicStore()
        store.addAsync(episodic(userText = "hi", assistantText = "hello"))
        val rag = RagPipelineBuilder.create().withStore(store).withTopK(2).withMaxCharsPerEntry(500).build()
        val ctx = rag.buildContextAsync("greeting")
        assertTrue(ctx.contains("hi"))
    }

    @Test
    fun `withInMemoryStore wires a fresh store`() = runTest {
        val rag = RagPipelineBuilder.create().withInMemoryStore().build()
        assertEquals("", rag.buildContextAsync("nothing stored"))
    }

    @Test
    fun `build without a store throws`() {
        assertFailsWith<IllegalStateException> { RagPipelineBuilder.create().build() }
    }

    @Test
    fun `withTopK rejects values below 1`() {
        assertFailsWith<IllegalArgumentException> { RagPipelineBuilder.create().withTopK(0) }
    }

    @Test
    fun `withMaxCharsPerEntry rejects values below 50`() {
        assertFailsWith<IllegalArgumentException> { RagPipelineBuilder.create().withMaxCharsPerEntry(49) }
    }

    @Test
    fun `withEmbedder wires the semantic-ranking seam`() = runTest {
        val store = InMemoryEpisodicStore()
        store.addAsync(episodic(userText = "near", assistantText = "n", embedding = floatArrayOf(1f, 0f)))
        store.addAsync(episodic(userText = "far", assistantText = "f", embedding = floatArrayOf(0f, 1f)))
        val embedder = object : ITextEmbedder {
            override suspend fun generateAsync(text: String): FloatArray = floatArrayOf(1f, 0f)
        }
        val rag = RagPipelineBuilder.create().withStore(store).withEmbedder(embedder).withTopK(1).build()
        val ctx = rag.buildContextAsync("q")
        assertTrue(ctx.contains("near"))
    }
}

File: typescript/tests/rag.test.ts
// rag.test.ts
//
// Exercises RagContextBuilder + RagPipelineBuilder. Mirrors
// CircleAI.Tests.RagContextBuilderTests plus the fluent-builder surface and the
// embedder ranking path.

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { RagContextBuilder, RagPipelineBuilder, type ITextEmbedder } from '../src/memory/rag';
import { InMemoryEpisodicStore } from '../src/memory/stores';
import type { EpisodicMemoryEntry, IEpisodicMemoryStore } from '../src/memory/index';

// ── Helpers ──────────────────────────────────────────────────────────────────

function episodic(overrides: Partial<EpisodicMemoryEntry>): EpisodicMemoryEntry {
  return {
    id: overrides.id ?? crypto.randomUUID(),
    recordedAtUtc: overrides.recordedAtUtc ?? new Date('2026-06-01T12:34:00Z'),
    userText: 'u',
    assistantText: 'a',
    ...overrides,
  };
}

function countOccurrences(text: string, token: string): number {
  let count = 0;
  let start = 0;
  for (;;) {
    const i = text.indexOf(token, start);
    if (i < 0) break;
    count++;
    start = i + token.length;
  }
  return count;
}

/** Store that always throws — used to test resilience. */
class ThrowingEpisodicStore implements IEpisodicMemoryStore {
  addAsync(): Promise<void> {
    throw new Error('store failure');
  }
  searchAsync(): Promise<readonly EpisodicMemoryEntry[]> {
    throw new Error('store failure');
  }
  getRecentAsync(): Promise<readonly EpisodicMemoryEntry[]> {
    throw new Error('store failure');
  }
  countAsync(): Promise<number> {
    throw new Error('store failure');
  }
  pruneOlderThanAsync(): Promise<number> {
    throw new Error('store failure');
  }
}

// ══════════════════════════════════════════════════════════════════════════
// Constructor guards
// ══════════════════════════════════════════════════════════════════════════

describe('RagContextBuilder — constructor guards', () => {
  it('throws when the store is null', () => {
    assert.throws(() => new RagContextBuilder(null as unknown as IEpisodicMemoryStore));
  });
});

// ══════════════════════════════════════════════════════════════════════════
// Empty / missing query
// ══════════════════════════════════════════════════════════════════════════

describe('RagContextBuilder — empty / missing query', () => {
  it('empty query returns empty', async () => {
    const b = new RagContextBuilder(new InMemoryEpisodicStore());
    assert.equal(await b.buildContextAsync(''), '');
  });

  it('whitespace query returns empty', async () => {
    const b = new RagContextBuilder(new InMemoryEpisodicStore());
    assert.equal(await b.buildContextAsync('   '), '');
  });
});

// ══════════════════════════════════════════════════════════════════════════
// Empty store
// ══════════════════════════════════════════════════════════════════════════

describe('RagContextBuilder — empty store', () => {
  it('empty store returns empty', async () => {
    const b = new RagContextBuilder(new InMemoryEpisodicStore());
    assert.equal(await b.buildContextAsync('hello'), '');
  });
});

// ══════════════════════════════════════════════════════════════════════════
// Non-empty store — recency fallback (no embedder)
// ══════════════════════════════════════════════════════════════════════════

describe('RagContextBuilder — formatting', () => {
  it('returns a formatted block with the header and both texts', async () => {
    const store = new InMemoryEpisodicStore();
    await store.addAsync(
      episodic({
        userText: 'What is SDPKT?',
        assistantText: 'SDPKT is the TGN wallet.',
        recordedAtUtc: new Date('2026-06-01T11:00:00Z'),
      }),
    );

    const b = new RagContextBuilder(store, null, 3);
    const result = await b.buildContextAsync('tell me about the wallet');

    assert.notEqual(result, '');
    assert.ok(result.includes('What is SDPKT?'));
    assert.ok(result.includes('SDPKT is the TGN wallet.'));
    assert.ok(result.includes('[Relevant past exchanges'));
  });

  it('formats the UTC timestamp as yyyy-MM-dd HH:mm and labels User/B!', async () => {
    const store = new InMemoryEpisodicStore();
    await store.addAsync(
      episodic({ userText: 'q', assistantText: 'r', recordedAtUtc: new Date('2026-06-01T09:05:00Z') }),
    );
    const b = new RagContextBuilder(store, null, 1);
    const result = await b.buildContextAsync('anything');
    assert.ok(result.includes('[2026-06-01 09:05 UTC]'));
    assert.ok(result.includes('User: q'));
    assert.ok(result.includes('B!: r'));
  });

  it('respects topK (counts bullet prefixes)', async () => {
    const store = new InMemoryEpisodicStore();
    for (let i = 0; i < 10; i++)
      await store.addAsync(episodic({ userText: `question ${i}`, assistantText: `answer ${i}` }));

    const b = new RagContextBuilder(store, null, 2);
    const result = await b.buildContextAsync('any question');
    assert.equal(countOccurrences(result, '• ['), 2);
  });

  it('includes the app context when set', async () => {
    const store = new InMemoryEpisodicStore();
    await store.addAsync(episodic({ userText: 'bid query', assistantText: 'bid answer', appContext: 'tgn.bidbaas' }));
    const b = new RagContextBuilder(store, null, 3);
    const result = await b.buildContextAsync('bidding');
    assert.ok(result.includes('tgn.bidbaas'));
  });

  it('truncates long texts to half the per-entry budget with an ellipsis', async () => {
    const store = new InMemoryEpisodicStore();
    const longText = 'x'.repeat(500);
    await store.addAsync(episodic({ userText: longText, assistantText: 'a' }));
    // maxCharsPerEntry 100 → half 50 → truncate to 49 chars + "…"
    const b = new RagContextBuilder(store, null, 1, 100);
    const result = await b.buildContextAsync('q');
    assert.ok(result.includes('x'.repeat(49) + '…'));
    assert.ok(!result.includes('x'.repeat(51)));
  });
});

// ══════════════════════════════════════════════════════════════════════════
// Embedder ranking path
// ══════════════════════════════════════════════════════════════════════════

describe('RagContextBuilder — embedder path', () => {
  it('ranks by the embedding when an embedder is supplied', async () => {
    const store = new InMemoryEpisodicStore();
    await store.addAsync(episodic({ userText: 'near', assistantText: 'n', embedding: [1, 0] }));
    await store.addAsync(episodic({ userText: 'far', assistantText: 'f', embedding: [0, 1] }));

    // Embedder maps any query to the x-axis, so "near" should rank first.
    const embedder: ITextEmbedder = { async generateAsync() { return [1, 0]; } };
    const b = new RagContextBuilder(store, embedder, 1);
    const result = await b.buildContextAsync('anything');
    assert.ok(result.includes('near'));
    assert.ok(!result.includes('far'));
  });

  it('falls back to recency when the embedder throws (still best-effort)', async () => {
    const store = new InMemoryEpisodicStore();
    await store.addAsync(
      episodic({ userText: 'only', assistantText: 'entry', recordedAtUtc: new Date('2026-06-01T00:00:00Z') }),
    );
    const embedder: ITextEmbedder = {
      async generateAsync() {
        throw new Error('embedder offline');
      },
    };
    const b = new RagContextBuilder(store, embedder, 3);
    const result = await b.buildContextAsync('q');
    assert.ok(result.includes('only'));
  });
});

// ══════════════════════════════════════════════════════════════════════════
// Resilience — store throws
// ══════════════════════════════════════════════════════════════════════════

describe('RagContextBuilder — resilience', () => {
  it('returns empty when the store throws (RAG is best-effort)', async () => {
    const b = new RagContextBuilder(new ThrowingEpisodicStore());
    assert.equal(await b.buildContextAsync('query'), '');
  });
});

// ══════════════════════════════════════════════════════════════════════════
// RagPipelineBuilder
// ══════════════════════════════════════════════════════════════════════════

describe('RagPipelineBuilder', () => {
  it('builds from an in-memory store and produces a working builder', async () => {
    const store = new InMemoryEpisodicStore();
    await store.addAsync(episodic({ userText: 'hi', assistantText: 'hello' }));
    const rag = RagPipelineBuilder.create().withStore(store).withTopK(2).withMaxCharsPerEntry(500).build();
    const ctx = await rag.buildContextAsync('greeting');
    assert.ok(ctx.includes('hi'));
  });

  it('withInMemoryStore wires a fresh store', async () => {
    const rag = RagPipelineBuilder.create().withInMemoryStore().build();
    assert.equal(await rag.buildContextAsync('nothing stored'), '');
  });

  it('build without a store throws', () => {
    assert.throws(() => RagPipelineBuilder.create().build(), /episodic memory store is required/i);
  });

  it('withTopK rejects values below 1', () => {
    assert.throws(() => RagPipelineBuilder.create().withTopK(0), RangeError);
  });

  it('withMaxCharsPerEntry rejects values below 50', () => {
    assert.throws(() => RagPipelineBuilder.create().withMaxCharsPerEntry(49), RangeError);
  });

  it('withEmbedder wires the semantic-ranking seam', async () => {
    const store = new InMemoryEpisodicStore();
    await store.addAsync(episodic({ userText: 'near', assistantText: 'n', embedding: [1, 0] }));
    await store.addAsync(episodic({ userText: 'far', assistantText: 'f', embedding: [0, 1] }));
    const embedder: ITextEmbedder = { async generateAsync() { return [1, 0]; } };
    const rag = RagPipelineBuilder.create().withStore(store).withEmbedder(embedder).withTopK(1).build();
    const ctx = await rag.buildContextAsync('q');
    assert.ok(ctx.includes('near'));
  });
});
